src/click/shell_completion.py

```diff
@@ -0,0 +1,465 @@
+COMPLETION_SCRIPT_BASH = """
+%(complete_func)s() {
+    local IFS=$'\n'
+    local response
+
+    response=$( env COMP_WORDS="${COMP_WORDS[*]}" \\
+                   COMP_CWORD=$COMP_CWORD \\
+                   %(autocomplete_var)s=complete $1 )
+
+    # if [[ -z $response ]]; then
+    #     COMPREPLY=""
+    # fi
+
+    for completion in $response; do
+        IFS=',' read type value <<< "$completion"
+        if [[ $type == 'dir' ]]; then
+            COMREPLY=()
+            compopt -o dirnames
+        elif [[ $type == 'file' ]]; then
+            COMREPLY=()
+            compopt -o default
+        elif [[ $type == 'none' ]]; then
+            COMPREPLY+=($value)
+        fi
+    done
+
+    return 0
+}
+
+%(complete_func)s_setup() {
+    complete -o nosort -F %(complete_func)s %(script_names)s
+}
+
+%(complete_func)s_setup
+"""
+
+COMPLETION_SCRIPT_ZSH = """
+#compdef %(script_names)s
+
+%(complete_func)s() {
+    local -a completions
+    local -a completions_with_descriptions
+    local -a response
+    (( ! $+commands[%(script_names)s] )) && return 1
+
+    response=("${(@f)$( env COMP_WORDS=\"${words[*]}\" \\
+                        COMP_CWORD=$((CURRENT-1)) \\
+                        %(autocomplete_var)s=\"complete_zsh\" \\
+                        %(script_names)s )}")
+
+    for key descr in ${(kv)response}; do
+      if [[ "$descr" == "_" ]]; then
+          completions+=("$key")
+      else
+          completions_with_descriptions+=("$key":"$descr")
+      fi
+    done
+
+    if [ -n "$completions_with_descriptions" ]; then
+        _describe -V unsorted completions_with_descriptions -U
+    fi
+
+    if [ -n "$completions" ]; then
+        compadd -U -V unsorted -a completions
+    fi
+    compstate[insert]="automenu"
+}
+
+compdef %(complete_func)s %(script_names)s
+"""
+
+
+COMPLETION_SCRIPT_FISH = """
+function %(complete_func)s_complete;
+    set -l response;
+    for value in (env %(autocomplete_var)s=complete_fish \
+        COMP_WORDS=(commandline -cp) COMP_CWORD=(commandline -t) \
+        %(script_names)s);
+        set response $response $value;
+    end;
+
+    for completion in $response;
+        set -l metadata (string split "," $completion);
+        if test $metadata[1] = "dir";
+            __fish_complete_directories $metadata[2];
+        else if test $metadata[1] = "file";
+            __fish_complete_path $metadata[2];
+        else if test $metadata[1] = "none";
+            echo $metadata[2];
+        end;
+    end;
+end;
+
+complete --no-files --command %(script_names)s --arguments \
+    "(%(complete_func)s_complete)"
+"""
+
+_invalid_ident_char_re = re.compile(r"[^a-zA-Z0-9_]")
+
+
+def start_of_option(param_str):
+    """Returns whether or not this si the start of an option declaration
+    (i.e, starts with "-" or "--")
+
+    :param param_str: param_str to check
+    """
+    return param_str and param_str[:1] == "-"
+
+
+class ShellComplete:
```

Comment:
Would this make more sense as an ABC? Really impressed with the PR so far.

Reply:
@kx-chen and I looked into ABC (as it's our first time using it), and we also thought it'll be a good idea

Reply:
seams like `ABC` is a good idea to force/indicate that the functions that currently `raise NotImplementedError` MUST be implemented.
and since they are not implemented in this `ShellComplete` its not usable as is, so its definitely abstract, isn't it?